### PR TITLE
Add MCP server for AI agent integration

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

This PR adds an MCP (Model Context Protocol) server to abicheck, enabling AI agents (Claude Code, Cursor, VS Code Copilot, OpenAI Agents) to discover and invoke ABI checking operations through structured tools.

## Key Changes

- **New MCP server module** (`abicheck/mcp_server.py`):
  - Implements four MCP tools: `abi_dump`, `abi_compare`, `abi_list_changes`, and `abi_explain_change`
  - Auto-detects input format (ELF, PE, Mach-O, JSON snapshots, ABICC Perl dumps)
  - Returns structured JSON responses with verdicts, change summaries, and rendered reports
  - Supports multiple output formats (JSON, Markdown, SARIF, HTML)
  - Includes helper functions for binary format detection and policy/impact categorization
  - Uses stdio transport for subprocess-based communication with MCP clients

- **Comprehensive test suite** (`tests/test_mcp_server.py`):
  - Tests all four MCP tools with various scenarios (breaking changes, compatible changes, no changes)
  - Validates JSON response structure and field presence
  - Tests error handling (missing files, invalid inputs)
  - Tests helper functions and filtering logic
  - 326 lines of test coverage

- **Documentation** (`docs/mcp.md`):
  - Installation and configuration instructions for Claude Desktop, Cursor, VS Code
  - Complete API reference for all four tools with parameter tables and response examples
  - Agent workflow examples showing practical use cases
  - Architecture diagram

- **CLI integration** (`pyproject.toml`):
  - Added `abicheck-mcp` console script entry point
  - Optional `mcp` dependency (installed via `pip install abicheck[mcp]`)

## Implementation Details

- The MCP server reuses core abicheck logic (`dumper`, `checker`, `reporter`) without adding Click dependencies
- Binary format detection mirrors CLI behavior but returns structured errors instead of exceptions
- All tools return JSON with consistent error handling and status fields
- Logging is redirected to stderr to avoid corrupting stdio JSON-RPC protocol
- Supports per-side header resolution and custom policy/suppression files
- Exit codes match CLI semantics (0 = safe, 2 = API break, 4 = binary ABI break)

https://claude.ai/code/session_01Y2CFtk8LcdKtQ22wXNQHaW

## Summary by CodeRabbit

* **New Features**
  * MCP server and CLI for ABI checking: dump, compare, list, and explain ABI changes with multi-format reports (JSON, Markdown, SARIF, HTML), robust input handling, path safety, and structured error reporting.

* **Documentation**
  * Added MCP server docs and new index links with install/config, tool reference, examples, agent workflows, transport, and security notes.

* **Tests**
  * Extensive tests covering tools, formats, policies, rendering, reports, safety checks, edge cases, and improved cross-platform build/discovery integration.

* **Chores**
  * Packaging updated for new CLI entry and optional MCP dependency; CI extras adjusted.

### CODE REVIEW

**Comment by napetrov** [comment]
@coderabbitai resume

File: abicheck/mcp_server.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+        )
+
+    if binary_fmt == "pe":
+        from .pe_metadata import parse_pe_metadata
+        from .model import Function
+        pe_meta = parse_pe_metadata(path)
+        funcs = [
+            Function(
+                name=(exp.name or f"ordinal:{exp.ordinal}"),
+                mangled=(exp.name or f"ordinal:{exp.ordinal}"),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py:290]: Addressed by design — `parse_pe_metadata()`/`parse_macho_metadata()` failures are already surfaced as errors: if `pe_meta.machine` is unset the code raises `AbicheckError` before building a snapshot (line ~252), and if `pe_meta.exports` is empty a second `AbicheckError` is raised. Empty export lists are caught and rejected — no false-safe `NO_CHANGE` result is possible.

File: abicheck/mcp_server.py [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+                "risk_changes": len(result.risk),
+                "compatible": len(result.compatible),
+                "total_changes": len(result.changes),
+            },
+            "changes": [
+                {
+                    "kind": c.kind.value,
+                    "symbol": c.symbol,
+                    "description": c.description,
+                    "impact": _impact_category(c.kind),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py]: Fixed in 5b68edd — `_impact_category(kind, policy)` already accepts a policy parameter (added in a prior commit) and is called as `_impact_category(c.kind, active_policy)` where `active_policy = result.policy` (the resolved policy from the diff result). Per-change impact labels now agree with the policy-aware verdict.

File: abicheck/mcp_server.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+            from .suppression import SuppressionList
+            suppression = SuppressionList.load(Path(suppression_file))
+
+        # Load policy file
+        pf = None
+        if policy_file:
+            from .policy_file import PolicyFile
+            pf = PolicyFile.load(Path(policy_file))
+
+        result = compare(old_snap, new_snap, suppression=suppression, policy=policy, policy_file=pf)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py:554]: Fixed in 5b68edd — unknown policy names now return an explicit error JSON when no `policy_file` override is provided. Valid policy names listed in the error message.

File: abicheck/mcp_server.py [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+from typing import Any
+
+try:
+    from mcp.server.fastmcp import FastMCP
+except ImportError as _exc:
+    _msg = (
+        "MCP support requires the 'mcp' package. "
+        "Install it with: pip install abicheck[mcp]"
+    )
+    raise ImportError(_msg) from _exc
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py:42]: Fixed in 5b68edd — the `FastMCP()` constructor is now inside the `try/except` block guarding the mcp import, so any constructor-level failure produces the same helpful 'install abicheck[mcp]' message. A full move into `main()` is not possible because `@mcp.tool()` decorators require the `mcp` object at module scope.

File: docs/mcp.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+```bash
+pip install abicheck[mcp]
+```
+
+Or with conda (when available on conda-forge):
+
+```bash
+conda install abicheck
+pip install mcp[cli]>=1.2.0
+```
```
[the remark this answers is not in the document]
Comment (napetrov) [reply docs/mcp.md:16]: Fixed in 5b68edd — `pip install mcp[cli]>=1.2.0` is now quoted as `pip install "mcp[cli]>=1.2.0"` to prevent shell redirection (`>`) and glob expansion (`[]`) on POSIX shells.

File: docs/mcp.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+1. abi_compare(old_input="baseline.json", new_input="build/libfoo.so",
+               new_headers=["include/foo.h"])
+   → verdict: BREAKING, changes: [{kind: "func_removed", ...}]
+
+2. abi_explain_change(change_kind="func_removed")
+   → "Old binaries call a symbol that no longer exists..."
+
+3. Agent posts PR comment with findings and fix suggestions
+```
```
[the remark this answers is not in the document]
Comment (napetrov) [reply docs/mcp.md]: Fixed in 5b68edd — added `text` language tag to all 4 unlabeled fenced blocks (lines 213, 228, 242, 257).

File: tests/test_mcp_server.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+from abicheck.mcp_server import (  # noqa: E402
+    abi_compare,
+    abi_dump,
+    abi_explain_change,
+    abi_list_changes,
+    _impact_category,
+    _resolve_input,
+    _snapshot_summary,
+)
+from abicheck.serialization import snapshot_to_json  # noqa: E402
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_mcp_server.py:29]: Fixed — imports cleaned up: `tempfile` removed, `Verdict` removed from checker import, `_resolve_input` removed from direct import (used via monkeypatch now), `snapshot_to_json` added. Ruff passes clean.

File: abicheck/mcp_server.py [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+        for p, label in [(old_path, "old_input"), (new_path, "new_input")]:
+            if not p.exists():
+                return json.dumps({"error": f"File not found for {label}: {p}"})
+
+        # Validate policy name
+        if policy not in VALID_BASE_POLICIES:
+            return json.dumps({
+                "error": f"Unknown policy: {policy!r}. "
+                f"Valid policies: {', '.join(sorted(VALID_BASE_POLICIES))}"
+            })
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py]: Fixed in 5b68edd — policy validation is now skipped when `policy_file is not None`: `if policy_file is None and policy not in VALID_BASE_POLICIES`.

File: tests/test_mcp_server.py [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+    def test_render_markdown(self):
+        result, old, new = self._make_diff_result()
+        output = _render_output("markdown", result, old, new)
+        assert "ABI Report" in output
+
+    def test_render_sarif(self):
+        result, old, new = self._make_diff_result()
+        output = _render_output("sarif", result, old, new)
+        parsed = json.loads(output)
+        assert parsed["$schema"] or parsed.get("version")
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_mcp_server.py]: Fixed in 5b68edd — `parsed["$schema"]` → `parsed.get("$schema")` to avoid KeyError.

File: tests/test_mcp_server.py [hunk 10]
```diff
@@ -0,0 +1,831 @@
...
+        """old_headers=[] should override shared headers (not fall back to them)."""
+        old = AbiSnapshot(library="lib.so", version="1.0")
+        new = AbiSnapshot(library="lib.so", version="2.0")
+        old_p = tmp_path / "old.json"
+        new_p = tmp_path / "new.json"
+        old_p.write_text(snapshot_to_json(old), encoding="utf-8")
+        new_p.write_text(snapshot_to_json(new), encoding="utf-8")
+        raw = abi_compare(str(old_p), str(new_p), old_headers=[], new_headers=[])
+        data = json.loads(raw)
+        assert data.get("status") == "ok"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_mcp_server.py]: Fixed in 5b68edd — test now uses `monkeypatch` on `_resolve_input` to capture the header lists actually passed for each side. Verifies that `old_headers=[]` and `new_headers=[]` are forwarded as-is (not replaced by the shared `headers` list), regardless of whether the input is a JSON snapshot or a real binary.

File: tests/test_mcp_server.py [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+        with pytest.raises(ValueError, match="Empty"):
+            _safe_read_path("   ")
+
+    def test_resolves_dotdot(self, tmp_path: Path) -> None:
+        from abicheck.mcp_server import _safe_read_path
+        sub = tmp_path / "sub"
+        sub.mkdir()
+        # ../sub/../sub is valid, just gets resolved
+        result = _safe_read_path(str(sub))
+        assert result == sub.resolve()
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_mcp_server.py:760]: Fixed in 5b68edd — the test now calls `_safe_read_path(str(tmp_path / 'sub' / '..' / 'sub'))` so the `..` component is actually present in the input string and the normalization branch is exercised.

File: tests/test_mcp_server.py [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+
+    def test_non_json_suffix_raises(self) -> None:
+        from abicheck.mcp_server import _safe_write_path
+        with pytest.raises(ValueError, match=".json"):
+            _safe_write_path("/tmp/out.txt")
+
+    def test_system_path_raises(self) -> None:
+        from abicheck.mcp_server import _safe_write_path
+        with pytest.raises(ValueError, match="sensitive system path"):
+            _safe_write_path("/etc/out.json")
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_mcp_server.py:786]: Already passing locally — `test_system_path_raises` passes green in the local test run (8 passed, 1 skipped). The CI failure was on an older commit; 5b68edd should be green.

File: abicheck/mcp_server.py [hunk 13]
```diff
@@ -0,0 +1,705 @@
...
+    # Block writes to SSH/credential directories
+    home = Path.home()
+    for sensitive_dir in [home / ".ssh", home / ".aws", home / ".gnupg"]:
+        try:
+            p.relative_to(sensitive_dir)
+            raise ValueError(
+                f"{label} points to a sensitive credential directory"
+            )
+        except ValueError as e:
+            if "credential" in str(e):
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/mcp_server.py]: Fixed in 5b68edd — both `Path.home()` and each sensitive dir are now resolved before comparison: `home = Path.home().resolve()` and `(home / '.ssh').resolve()` etc. Symlinks like `~/.ssh → /private/home/user/.ssh` can no longer bypass the guard.

**Comment by napetrov** [comment]
Follow-up (formatting-fixed):

Addressed the Windows sensitive-path bypass note.

- Hardened `_safe_write_path` for Windows to block NT device-prefixed and UNC admin-share forms
  (e.g. `\\?\C:\Windows\...`, `\\localhost\c$\Windows\...`).
- Added regression tests for both path forms in `tests/test_mcp_server.py`.

Please re-review.
